ext/drpttracer/drpttracer.dox

```diff
@@ -0,0 +1,130 @@
+
+ - \ref sec_drpttracer_setup
+ - \ref sec_drpttracer_usage
+ - \ref sec_drpttracer_tracing_mode
+
+\section sec_drpttracer_setup Setup
+
+To use \p drpttracer with your client simply include this line in your client's
+\p CMakeLists.txt file:
+
+\code use_DynamoRIO_extension(clientname drpttracer) \endcode
+
+That will automatically set up the include path and library dependence.
+
+Initialize and clean up \p drpttracer by calling drpttracer_init() and
+drpttracer_exit().
+
+\section sec_drpttracer_usage Usage
+
+Intel PT (Processor Trace) only logs control flow changes. Therefore, when decoding a
+trace, the decoder needs to get raw bits of every instruction from images.
+\p drpttracer provides APIs to use Intel PT to generate trace data and some auxiliary data
+to help the decoder (e.g. libipt) to decode the trace. One type of auxiliary data is
+sideband data. This sideband data stores perf event records, which contain the image
+change message necessary for decoding the trace. Another type of auxiliary data is the
+metadata for a trace. The metadata contains the CPU information and some other information
+that can be used to synthesize the time of PT trace and the perf event records. The
+auxiliary data enables the decoder to find the correct raw bits for every instruction.
+
+\p drpttracer provides one data structure to store all generated data:
+
+ - #drpttracer_output_t
+
+And \p drpttracer provides two APIs for client to start and stop tracing:
+
+ - drpttracer_start_tracing()
+ - drpttracer_end_tracing()
+
+To record a PT trace, the client uses drpttracer_start_tracing() to start the tracing. The
+start function lets the client specify the following parameters:
+
+ - trace_mode : The tracing mode.
+
+ - pt_size_shift: The size shift of PT trace's ring buffer. It must be greater than 0, and
+the buffer size is 2^pt_size_shift * PAGE_SIZE. The client nees to make sure the buffer
+size is large enough to store the trace.
+
+ - sideband_size_shift: The size shift of PT sideband data's ring buffer. It must be
+greater than 0, and the buffer size is 2^sideband_size_shift * PAGE_SIZE. The client nees
+to make sure the buffer size is large enough to store the sideband data.
+
+And the start function will generate a handle for the PT tracing. The handle is a pointer
+to a \p drpttracer's internal data structure which contains the tracing's PT data buffer,
+sideband data buffer, and metadata.
+
+To stop collecting the trace, the client needs to pass the trace handle to
+drpttracer_end_tracing(). When the tracing is stopped, the stop function will allocate an
+instance of #drpttracer_output_t, and copy the PT data, sideband data, and metadata to an
+#drpttracer_output_t instance. Then return the instance to the client. The client can dump
+the data in the output to different files for offline post-processing. And the user can
+use the library drpt2ir in drcachesim to convert the PT trace to DynamoRIO's IR.
```

Comment:
We will not build drpt2ir's doc in arm or x86_32 platforms. So do I not build drpttracer.dox on Arm or x86_32?

Reply:
Generally we want to build all docs on all platforms.  Think about the single copy of the docs at dynamorio.org: that should cover all platforms (with notes in the individual doc entries about which platforms are supported) -- we're not going to generate 5 different copies for the online docs just because a few components are limited to certain platforms.

The docs cmake code looks in the source dirs for ext/*/*.dox so are you sure the docs aren't built on all platforms today?

Reply:
On that note, please check drpttracer.dox and make sure it says up front what platforms it supports.  (Maybe every function should also have a short note too??  Open to debate.)

Reply:
So, is the plan that we build drpttracer and pt2ir on all platforms, but in drpttracer_init etc we return false for non-x86? That's reasonable.

Qidong and I spoke offline yesterday about fixing some of this build stuff in a follow up PR, We can perhaps let the drpt2ir be as-is for now, and we can fix it in the same follow-up PR?

Reply:
There's no need to build the libraries on all platforms: just the docs.  And as I pointed out above that should already be happening.

Reply:
> And as I pointed out above that should already be happening.

For ext/ anyway via globs in the docs code: but drpt2ir is not in ext/ -- so that's not being built?  Sounds like the action item is to find the dr$sim doxygen globs and update those.

Reply:
> We can perhaps let the drpt2ir reference in this dox be as-is for now, and we can fix it in the same follow-up PR?

OK.